packages/flutter/lib/src/foundation/persistent_hash_map.dart

```diff
@@ -0,0 +1,401 @@
+///  * [Clojure's `PersistentHashMap`](https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/PersistentHashMap.java).
+///
+class PersistentHashMap<K extends Object, V> {
+  /// An an empty hash map.
+  const PersistentHashMap.empty() : this._(null);
+
+  const PersistentHashMap._(this._root);
+
+  final _TrieNode? _root;
+
+  /// If this map does not already contain the given [key] to [value]
+  /// mapping then create a new version of the map which contains
+  /// all mappings from the current one plus the given [key] to [value]
+  /// mapping.
+  PersistentHashMap<K, V> put(K key, V value) {
+    final _TrieNode newRoot =
+        (_root ?? _CompressedNode.empty).put(0, key, key.hashCode, value);
+    if (newRoot == _root) {
+      return this;
+    }
+    return PersistentHashMap<K, V>._(newRoot);
+  }
+
+  /// Returns value associated with the given [key] or `null` if [key]
+  /// is not in the map.
+  @pragma('dart2js:as:trust')
+  V? operator[](K key) {
+    if (_root == null) {
+      return null;
+    }
+
+    // Unfortunately can not use unsafeCast<V?>(...) here because it leads
+    // to worse code generation on VM.
+    return _root!.get(0, key, key.hashCode) as V?;
+  }
+}
+
+/// Base class for nodes in a hash trie.
+///
+/// This trie is keyed by hash code bits using [hashBitsPerLevel] bits
+/// at each level.
+abstract class _TrieNode {
+  static const int hashBitsPerLevel = 5;
+  static const int hashBitsPerLevelMask = (1 << hashBitsPerLevel) - 1;
+
+  @pragma('vm:prefer-inline')
+  static int trieIndex(int hash, int bitIndex) {
+    return (hash >>> bitIndex) & hashBitsPerLevelMask;
+  }
+
+  /// Insert [key] to [value] mapping into the trie using bits from [keyHash]
+  /// starting at [bitIndex].
+  _TrieNode put(int bitIndex, Object key, int keyHash, Object? value);
+
+  /// Lookup a value associated with the given [key] using bits from [keyHash]
+  /// starting at [bitIndex].
+  Object? get(int bitIndex, Object key, int keyHash);
+}
+
+/// A full (uncompressed) node in the trie.
+///
+/// It contains an array with `1<<_hashBitsPerLevel` elements which
+/// are references to deeper nodes.
+class _FullNode extends _TrieNode {
+  _FullNode(this.descendants);
+
+  static const int numElements = 1 << _TrieNode.hashBitsPerLevel;
+
+  // Caveat: this array is actually List<_TrieNode?> but typing it like that
+  // will introduce a type check when copying this array. For performance
+  // reasons we instead omit the type and use (implicit) casts when accessing
+  // it instead.
+  final List<Object?> descendants;
+
+  @override
+  _TrieNode put(int bitIndex, Object key, int keyHash, Object? value) {
+    final int index = _TrieNode.trieIndex(keyHash, bitIndex);
+    final _TrieNode node = _unsafeCast<_TrieNode?>(descendants[index]) ?? _CompressedNode.empty;
+    final _TrieNode newNode = node.put(
+        bitIndex + _TrieNode.hashBitsPerLevel, key, keyHash, value);
```

Comment:
nit: (here and elsewhere): Flutter style wants the line of the opening and closing parenthesis to have to same indentation. Either put this all on one line or format it with a trailing comma (each param on a separate line followed by the parenthesis on the next line.